ppl/lake/walk.go

```diff
@@ -111,32 +114,40 @@ func tsDirEntriesToChunks(ctx context.Context, lk *Lake, filterSpan nano.Span, t
 			m[id] = s
 		}
 	}
-	var chunks []chunk.Chunk
+
+	var mu sync.Mutex
+	chunks := make([]chunk.Chunk, 0, len(m))
+	group, ctx := errgroup.WithContext(ctx)
+
 	for id, seen := range m {
 		if !seen.meta || !seen.data {
 			continue
 		}
-		dir := tsDir.path(lk)
-		mdPath := chunk.MetadataPath(dir, id)
-		b, err := lk.immfiles.ReadFile(ctx, mdPath)
-		if err != nil {
-			return nil, err
-		}
+		id := id
+		group.Go(func() error {
+			dir := tsDir.path(lk)
+			mdPath := chunk.MetadataPath(dir, id)
+			b, err := lk.immfiles.ReadFile(ctx, mdPath)
+			if err != nil {
+				if zqe.IsNotFound(err) {
+					return nil
+				}
+				return fmt.Errorf("failed to read chunk metadata from %v: %w", mdPath, err)
+			}
 
-		md, err := chunk.UnmarshalMetadata(b, lk.DataOrder)
-		if err != nil {
-			if zqe.IsNotFound(err) {
-				continue
+			md, err := chunk.UnmarshalMetadata(b, lk.DataOrder)
+			if err != nil {
+				return err
 			}
-			return nil, fmt.Errorf("failed to read chunk metadata from %v: %w", mdPath, err)
-		}
-		chunk := md.Chunk(dir, id)
-		if !filterSpan.Overlaps(chunk.Span()) {
-			continue
-		}
```

Comment:
Nit: I think checking for overlap here rather than in the tsDirEntriesToChunks caller seems like a better approach, mostly because it's less code.

Reply:
I think it's easier to reason about tsDirEntriesToChunks if it literally just returns the chunks in a tsdir. Passing in filterSpan felt like cramming too much functionality into this function. Plus I consider any functionality we can move out walk.go and into other packages a win.

Reply:
Fair enough!